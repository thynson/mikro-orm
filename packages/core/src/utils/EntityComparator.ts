import { AnyEntity, Dictionary, EntityData, EntityMetadata, EntityProperty, HelperType, IMetadataStorage } from '../typings';
import { ReferenceType } from '../enums';
import { Platform } from '../platforms';
import { Utils } from './Utils';

export class EntityComparator {

  constructor(private readonly metadata: IMetadataStorage,
              private readonly platform: Platform) { }

  /**
   * Computes difference between two entities. First calls `prepareEntity` on both, then uses the `diff` method.
   */
  diffEntities<T extends AnyEntity<T>>(a: T, b: T): EntityData<T> {
    return Utils.diff(this.prepareEntity(a), this.prepareEntity(b)) as EntityData<T>;
  }

  /**
   * Removes ORM specific code from entities and prepares it for serializing. Used before change set computation.
   * References will be mapped to primary keys, collections to arrays of primary keys.
   */
  prepareEntity<T extends AnyEntity<T>>(entity: T): EntityData<T> {
    if ((entity as Dictionary).__prepared) {
      return entity;
    }

    const meta = this.metadata.get<T>(entity.constructor.name);
    const ret = {} as EntityData<T>;

    if (meta.discriminatorValue) {
      ret[meta.root.discriminatorColumn as keyof T] = meta.discriminatorValue as unknown as T[keyof T];
    }

    // copy all comparable props, ignore collections and references, process custom types
    meta.comparableProps.forEach(prop => {
      if (this.shouldIgnoreProperty(entity, prop)) {
        return;
      }

      if (prop.reference === ReferenceType.EMBEDDED) {
        return meta.props.filter(p => p.embedded?.[0] === prop.name).forEach(childProp => {
          ret[childProp.name as keyof T] = Utils.copy(entity[prop.name][childProp.embedded![1]]);
        });
      }

      if (Utils.isEntity(entity[prop.name], true)) {
        ret[prop.name] = Utils.getPrimaryKeyValues(entity[prop.name], this.metadata.find(prop.type)!.primaryKeys, true);

        if (prop.customType) {
          return ret[prop.name] = Utils.copy(prop.customType.convertToDatabaseValue(ret[prop.name], this.platform));
        }

        return;
      }

      if (prop.customType) {
        return ret[prop.name] = Utils.copy(prop.customType.convertToDatabaseValue(entity[prop.name], this.platform));
      }

      if (prop.type.toLowerCase() === 'date') {
        return ret[prop.name] = Utils.copy(this.platform.processDateProperty(entity[prop.name]));
      }

      if (Array.isArray(entity[prop.name]) || Utils.isObject(entity[prop.name])) {
        return ret[prop.name] = Utils.copy(entity[prop.name]);
      }

      ret[prop.name] = Utils.copy(entity[prop.name]);
    });

    Object.defineProperty(ret, '__prepared', { value: true });

    return ret;
  }

  /**
   * should be used only for `meta.comparableProps` that are defined based on the static `isComparable` helper
   */
  private shouldIgnoreProperty<T extends AnyEntity<T>>(entity: T, prop: EntityProperty<T>) {
    if (!(prop.name in entity)) {
      return true;
    }

    const value = entity[prop.name];
    const noPkRef = Utils.isEntity<T>(value, true) && !value[HelperType]!.hasPrimaryKey();
    const noPkProp = prop.primary && !Utils.isDefined(value, true);

    // bidirectional 1:1 and m:1 fields are defined as setters, we need to check for `undefined` explicitly
    const isSetter = [ReferenceType.ONE_TO_ONE, ReferenceType.MANY_TO_ONE].includes(prop.reference) && (prop.inversedBy || prop.mappedBy);
    const emptyRef = isSetter && value === undefined;

    return noPkProp || noPkRef || emptyRef || prop.version;
  }

  /**
   * perf: used to generate list of comparable properties during discovery, so we speed up the runtime comparison
   */
  static isComparable<T extends AnyEntity<T>>(prop: EntityProperty<T>, root: EntityMetadata) {
    const virtual = prop.persist === false;
    const inverse = prop.reference === ReferenceType.ONE_TO_ONE && !prop.owner;
    const discriminator = prop.name === root.discriminatorColumn;
    const collection = prop.reference === ReferenceType.ONE_TO_MANY || prop.reference === ReferenceType.MANY_TO_MANY;

    return !virtual && !collection && !inverse && !discriminator && !prop.version;
  }

}
