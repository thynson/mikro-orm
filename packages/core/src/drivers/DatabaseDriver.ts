import { EntityManagerType, FindOneOptions, FindOptions, IDatabaseDriver } from './IDatabaseDriver';
import { EntityData, EntityMetadata, EntityProperty, FilterQuery, AnyEntity, Dictionary, Primary, PopulateOptions, HelperType } from '../typings';
import { MetadataStorage } from '../metadata';
import { Connection, QueryResult, Transaction } from '../connections';
import { Configuration, ConnectionOptions, Utils } from '../utils';
import { LockMode, QueryOrder, QueryOrderMap, ReferenceType } from '../enums';
import { Platform } from '../platforms';
import { Collection } from '../entity';
import { EntityManager } from '../EntityManager';
import { ValidationError } from '../errors';
import { DriverException } from '../exceptions';

export abstract class DatabaseDriver<C extends Connection> implements IDatabaseDriver<C> {

  [EntityManagerType]: EntityManager<this>;

  protected readonly connection!: C;
  protected readonly replicas: C[] = [];
  protected readonly platform!: Platform;
  protected readonly logger = this.config.getLogger();
  protected metadata!: MetadataStorage;

  protected constructor(protected readonly config: Configuration,
                        protected readonly dependencies: string[]) { }

  abstract async find<T extends AnyEntity<T>>(entityName: string, where: FilterQuery<T>, options?: FindOptions<T>, ctx?: Transaction): Promise<T[]>;

  abstract async findOne<T extends AnyEntity<T>>(entityName: string, where: FilterQuery<T>, options?: FindOneOptions<T>, ctx?: Transaction): Promise<T | null>;

  abstract async nativeInsert<T extends AnyEntity<T>>(entityName: string, data: EntityData<T>, ctx?: Transaction): Promise<QueryResult>;

  abstract async nativeInsertMany<T extends AnyEntity<T>>(entityName: string, data: EntityData<T>[], ctx?: Transaction): Promise<QueryResult>;

  abstract async nativeUpdate<T extends AnyEntity<T>>(entityName: string, where: FilterQuery<T>, data: EntityData<T>, ctx?: Transaction): Promise<QueryResult>;

  abstract async nativeDelete<T extends AnyEntity<T>>(entityName: string, where: FilterQuery<T>, ctx?: Transaction): Promise<QueryResult>;

  abstract async count<T extends AnyEntity<T>>(entityName: string, where: FilterQuery<T>, ctx?: Transaction): Promise<number>;

  createEntityManager<D extends IDatabaseDriver = IDatabaseDriver>(useContext?: boolean): D[typeof EntityManagerType] {
    return new EntityManager(this.config, this, this.metadata, useContext) as unknown as EntityManager<D>;
  }

  async aggregate(entityName: string, pipeline: any[]): Promise<any[]> {
    throw new Error(`Aggregations are not supported by ${this.constructor.name} driver`);
  }

  async loadFromPivotTable<T extends AnyEntity<T>, O extends AnyEntity<O>>(prop: EntityProperty, owners: Primary<O>[][], where?: FilterQuery<T>, orderBy?: QueryOrderMap, ctx?: Transaction): Promise<Dictionary<T[]>> {
    throw new Error(`${this.constructor.name} does not use pivot tables`);
  }

  async syncCollection<T extends AnyEntity<T>, O extends AnyEntity<O>>(coll: Collection<T, O>, ctx?: Transaction): Promise<void> {
    const pk = this.metadata.find(coll.property.type)!.primaryKeys[0];
    const data = { [coll.property.name]: coll.getIdentifiers(pk) } as EntityData<T>;
    await this.nativeUpdate<T>(coll.owner.constructor.name, coll.owner[HelperType]!.__primaryKey, data, ctx);
  }

  mapResult<T extends AnyEntity<T>>(result: EntityData<T>, meta: EntityMetadata, populate: PopulateOptions<T>[] = []): T | null {
    if (!result || !meta) {
      return null;
    }

    const ret = Object.assign({}, result) as any;

    meta.props.forEach(prop => {
      if (prop.fieldNames && prop.fieldNames.length > 1 && prop.fieldNames.every(joinColumn => Utils.isDefined(ret[joinColumn], true))) {
        const temp: any[] = [];
        prop.fieldNames.forEach(joinColumn => {
          temp.push(ret[joinColumn]);
          delete ret[joinColumn];
        });

        ret[prop.name] = temp;
      } else if (prop.fieldNames && prop.fieldNames[0] in ret) {
        Utils.renameKey(ret, prop.fieldNames[0], prop.name);
      }

      if (prop.type === 'boolean' && ![null, undefined].includes(ret[prop.name])) {
        ret[prop.name] = !!ret[prop.name];
      }
    });

    return ret as T;
  }

  async connect(): Promise<C> {
    await this.connection.connect();
    await Promise.all(this.replicas.map(replica => replica.connect()));

    return this.connection;
  }

  async reconnect(): Promise<C> {
    await this.close(true);
    return this.connect();
  }

  getConnection(type: 'read' | 'write' = 'write'): C {
    if (type === 'write' || this.replicas.length === 0) {
      return this.connection as C;
    }

    const rand = Utils.randomInt(0, this.replicas.length - 1);

    return this.replicas[rand] as C;
  }

  async close(force?: boolean): Promise<void> {
    await Promise.all(this.replicas.map(replica => replica.close(force)));
    await this.connection.close(force);
  }

  getPlatform(): Platform {
    return this.platform;
  }

  setMetadata(metadata: MetadataStorage): void {
    this.metadata = metadata;
    this.connection.setMetadata(metadata);
  }

  getDependencies(): string[] {
    return this.dependencies;
  }

  async ensureIndexes(): Promise<void> {
    throw new Error(`${this.constructor.name} does not use ensureIndexes`);
  }

  protected inlineEmbeddables<T>(meta: EntityMetadata<T>, data: T): void {
    Object.keys(data).forEach(k => {
      if (Utils.isOperator(k)) {
        Utils.asArray(data[k]).forEach(payload => this.inlineEmbeddables(meta, payload));
      }
    });

    meta.props.forEach(prop => {
      if (prop.reference === ReferenceType.EMBEDDED && Utils.isObject(data[prop.name])) {
        const props = prop.embeddedProps;

        Object.keys(data[prop.name]).forEach(kk => {
          const operator = Object.keys(data[prop.name]).some(f => Utils.isOperator(f));

          if (operator) {
            throw ValidationError.cannotUseOperatorsInsideEmbeddables(meta.name!, prop.name, data);
          }

          data[props[kk].name] = data[prop.name][props[kk].embedded![1]];
        });
        delete data[prop.name];
      }
    });
  }

  protected getPivotOrderBy(prop: EntityProperty, orderBy?: QueryOrderMap): QueryOrderMap {
    if (orderBy) {
      return orderBy;
    }

    if (prop.orderBy) {
      return prop.orderBy;
    }

    if (prop.fixedOrder) {
      return { [`${prop.pivotTable}.${prop.fixedOrderColumn}`]: QueryOrder.ASC };
    }

    return {};
  }

  protected getPrimaryKeyFields(entityName: string): string[] {
    const meta = this.metadata.find(entityName);
    return meta ? meta.primaryKeys : [this.config.getNamingStrategy().referenceColumnName()];
  }

  protected getPivotInverseProperty(prop: EntityProperty): EntityProperty {
    const pivotMeta = this.metadata.find(prop.pivotTable)!;
    let inverse: string;

    if (prop.owner) {
      const pivotProp1 = pivotMeta.properties[prop.type + '_inverse'];
      inverse = pivotProp1.mappedBy;
    } else {
      const pivotProp1 = pivotMeta.properties[prop.type + '_owner'];
      inverse = pivotProp1.inversedBy;
    }

    return pivotMeta.properties[inverse];
  }

  protected createReplicas(cb: (c: ConnectionOptions) => C): C[] {
    const replicas = this.config.get('replicas', [])!;
    const ret: C[] = [];
    const props = ['dbName', 'clientUrl', 'host', 'port', 'user', 'password', 'multipleStatements', 'pool', 'name'] as const;

    replicas.forEach((conf: Partial<ConnectionOptions>) => {
      props.forEach(prop => (conf[prop] as any) = prop in conf ? conf[prop] : this.config.get(prop));
      ret.push(cb(conf as ConnectionOptions));
    });

    return ret;
  }

  async lockPessimistic<T extends AnyEntity<T>>(entity: T, mode: LockMode, ctx?: Transaction): Promise<void> {
    throw new Error(`Pessimistic locks are not supported by ${this.constructor.name} driver`);
  }

  protected shouldHaveColumn<T extends AnyEntity<T>>(prop: EntityProperty<T>, populate: PopulateOptions<T>[], includeFormulas = true): boolean {
    if (prop.formula) {
      return includeFormulas;
    }

    if (prop.persist === false) {
      return false;
    }

    if (prop.lazy && !populate.some(p => p.field === prop.name)) {
      return false;
    }

    return [ReferenceType.SCALAR, ReferenceType.MANY_TO_ONE].includes(prop.reference) || (prop.reference === ReferenceType.ONE_TO_ONE && prop.owner);
  }

  /**
   * @inheritDoc
   */
  convertException(exception: Error): DriverException {
    if (exception instanceof DriverException) {
      return exception;
    }

    return this.platform.getExceptionConverter().convertException(exception);
  }

  protected rethrow<T>(promise: Promise<T>): Promise<T> {
    return promise.catch(e => {
      throw this.convertException(e);
    });
  }

}
