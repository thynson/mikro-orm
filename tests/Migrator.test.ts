import umzug from 'umzug';

import { unlink } from 'fs-extra';
import { initORMMySql } from './bootstrap';
import { Configuration, Logger } from '../lib/utils';
import { Migration, Migrator, MikroORM } from '../lib';
import { MongoDriver } from '../lib/drivers/MongoDriver';
import { MySqlDriver } from '../lib/drivers/MySqlDriver';

class MigrationTest1 extends Migration {

  async up(): Promise<void> {
    this.addSql('select 1 + 1');
  }

}

class MigrationTest2 extends Migration {

  async up(): Promise<void> {
    this.addSql('select 1 + 1');
  }

  isTransactional(): boolean {
    return false;
  }

}

describe('Migrator', () => {

  let orm: MikroORM<MySqlDriver>;

  beforeAll(async () => orm = await initORMMySql());
  afterAll(async () => orm.close(true));

  test('generate schema migration', async () => {
    const dateMock = jest.spyOn(Date.prototype, 'toISOString');
    dateMock.mockReturnValue('2019-10-13T21:48:13.382Z');
    const migrator = orm.getMigrator();
    const migration = await migrator.createMigration();
    expect(migration).toMatchSnapshot('migration-dump');
    await unlink(process.cwd() + '/temp/migrations/Migration20191013214813.ts');
  });

  test('run schema migration', async () => {
    const upMock = jest.spyOn(umzug.prototype, 'up');
    const downMock = jest.spyOn(umzug.prototype, 'down');
    upMock.mockImplementationOnce(() => {});
    downMock.mockImplementationOnce(() => {});
    const migrator = orm.getMigrator();
    await migrator.up();
    expect(upMock).toBeCalledTimes(1);
    expect(downMock).toBeCalledTimes(0);
    await migrator.down();
    expect(upMock).toBeCalledTimes(1);
    expect(downMock).toBeCalledTimes(1);
    upMock.mockRestore();
  });

  test('ensureTable and list executed migrations', async () => {
    await orm.em.getConnection().getKnex().schema.dropTableIfExists(orm.config.get('migrations').tableName!);
    const migrator = orm.getMigrator();
    // @ts-ignore
    const storage = migrator.storage;

    await storage.ensureTable(); // creates the table
    await storage.logMigration('test');
    await expect(storage.getExecutedMigrations()).resolves.toMatchObject([{ name: 'test' }]);
    await expect(storage.executed()).resolves.toEqual(['test']);

    await storage.ensureTable(); // table exists, no-op
    await storage.unlogMigration('test');
    await expect(storage.executed()).resolves.toEqual([]);

    await expect(migrator.getPendingMigrations()).resolves.toEqual([]);
  });

  test('runner', async () => {
    await orm.em.getConnection().getKnex().schema.dropTableIfExists(orm.config.get('migrations').tableName!);
    const migrator = orm.getMigrator();
    // @ts-ignore
    await migrator.storage.ensureTable();
    // @ts-ignore
    const runner = migrator.runner;

    const mock = jest.fn();
    const logger = new Logger(mock, true);
    Object.assign(orm.config, { logger });

    const migration1 = new MigrationTest1(orm.em.getDriver(), orm.config);
    const spy1 = jest.spyOn(Migration.prototype, 'addSql');
    mock.mock.calls.length = 0;
    await runner.run(migration1, 'up');
    expect(spy1).toBeCalledWith('select 1 + 1');
    expect(mock.mock.calls.length).toBe(6);
    expect(mock.mock.calls[0][0]).toMatch('begin');
    expect(mock.mock.calls[1][0]).toMatch('set names utf8;');
    expect(mock.mock.calls[2][0]).toMatch('set foreign_key_checks = 0;');
    expect(mock.mock.calls[3][0]).toMatch('select 1 + 1');
    expect(mock.mock.calls[4][0]).toMatch('set foreign_key_checks = 1;');
    expect(mock.mock.calls[5][0]).toMatch('commit');
    mock.mock.calls.length = 0;

    await expect(runner.run(migration1, 'down')).rejects.toThrowError('This migration cannot be reverted');
    const executed = await migrator.getExecutedMigrations();
    expect(executed).toEqual([]);

    mock.mock.calls.length = 0;
    // @ts-ignore
    migrator.options.disableForeignKeys = false;
    const migration2 = new MigrationTest2(orm.em.getDriver(), orm.config);
    await runner.run(migration2, 'up');
    expect(mock.mock.calls.length).toBe(1);
    expect(mock.mock.calls[0][0]).toMatch('select 1 + 1');
  });

  test('resolver', async () => {
    // TODO mock umzug to load the MigrationTest1 and trigger the resolver
    await orm.em.getConnection().getKnex().schema.dropTableIfExists(orm.config.get('migrations').tableName!);
    jest.mock('../../temp/migrations/MigrationTest1.ts', () => ({ MigrationTest1 }), { virtual: true });
    const migrator = orm.getMigrator();
    await migrator.up();
  });

  test('not supported [mongodb]', async () => {
    const mongoOrm = Object.create(MikroORM.prototype, { driver: new MongoDriver(new Configuration({} as any, false)) } as any);
    expect(() => mongoOrm.getMigrator()).toThrowError('Not supported by given driver');
  });

});
