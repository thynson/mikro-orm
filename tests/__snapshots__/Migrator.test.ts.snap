// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Migrator generate schema migration: migration-dump 1`] = `
Array [
  "import { Migration } from 'mikro-orm';

export class Migration20191013214813 extends Migration {

  async up(): Promise<void> {
    this.addSql('alter table \`book2\` drop \`foo\`;');

    this.addSql('alter table \`test2\` drop foreign key \`test2_foo___bar_foreign\`;');
    this.addSql('alter table \`test2\` drop index \`test2_foo___bar_unique\`;');
    this.addSql('alter table \`test2\` drop index \`test2_foo___bar_index\`;');
    this.addSql('alter table \`test2\` drop \`foo___bar\`;');
    this.addSql('alter table \`test2\` drop \`foo___baz\`;');
  }

}
",
  Array [
    "alter table \`book2\` drop \`foo\`;",
    "",
    "alter table \`test2\` drop foreign key \`test2_foo___bar_foreign\`;",
    "alter table \`test2\` drop index \`test2_foo___bar_unique\`;",
    "alter table \`test2\` drop index \`test2_foo___bar_index\`;",
    "alter table \`test2\` drop \`foo___bar\`;",
    "alter table \`test2\` drop \`foo___baz\`;",
    undefined,
    undefined,
  ],
]
`;
